src/cargo/ops/tree/graph.rs

```diff
@@ -391,10 +391,20 @@ fn add_pkg(
         let dep_pkg = graph.package_map[&dep_id];
 
         for dep in deps {
-            let dep_features_for = if dep.is_build() || dep_pkg.proc_macro() {
-                FeaturesFor::HostDep
-            } else {
-                features_for
+            let dep_features_for = match dep
```

Comment:
Thanks for reviving this PR! 

However, this is missing an important part of [my original patch](https://github.com/rust-lang/cargo/issues/10593#issuecomment-1317759526). Please see my [previous comment](https://github.com/rust-lang/cargo/pull/13207#discussion_r1543502588) :)

I am tempted to merge this PR as it fixes some panics. I will be more existed if we can implement the design proposed here: <https://github.com/rust-lang/cargo/issues/10593#issuecomment-1341214604>.

Reply:
I'm very happy to keep implementing the design! (I'm now emotionally invested), but I think we should split it into separate PRs.'ve updated the PR message to explain each commit as they currently exist.

Reply:
Update: Moved the proposed design change work to #14658 to keep this PR clear.